File: Coding_1006 Binary_Search_And_Linked_List/최원준/이강훈/3번.java
/*
https://leetcode.com/problems/median-of-two-sorted-arrays/

아이디어
원준이형 강의 링크 봤는데 구현이 안됨 실패

자료구조

시간복잡도

*/

public class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int total = nums1.length + nums2.length;
        int half = total / 2;

        if (nums1.length > nums2.length) {
            int[] temp = nums1.clone();
            nums1 = nums2;
            nums2 = temp;
        }

        int l = 0;
        int r = nums1.length - 1;
        int m = (l + r) / 2;
        int nums2Position = half - (m + 2);
        int nums1Next = m < nums1.length - 1 ? nums1[m + 1] : Integer.MAX_VALUE;
        int nums2Next = nums2Position < nums2.length - 1 ? nums2[nums2Position + 1] : Integer.MAX_VALUE;
        while (l <= r) {
            if (nums1[m] > nums2Next) r = m - 1;
            else if (nums2[nums2Position] > nums1Next) l = m + 1;
            else break;

            m = (l + r) / 2;
            nums2Position = half - (m + 2);

            nums1Next = m < nums1.length - 1 ? nums1[m + 1] : Integer.MAX_VALUE;
            nums2Next = nums2Position < nums2.length - 1 ? nums2[nums2Position + 1] : Integer.MAX_VALUE;
        }

        return total % 2 == 0
                ? (double) (Math.max(nums1[m], nums2[nums2Position]) + Math.min(nums1Next, nums2Next)) / 2
                : Math.min(nums1Next, nums2Next);
    }
}

File: Coding_1006 Binary_Search_And_Linked_List/최원준/최원준/3번.py
# https://leetcode.com/problems/median-of-two-sorted-arrays/description/

'''
1. 아이디어 :

2. 시간복잡도 :

3. 자료구조 :

'''

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        #작은 배열을 바꾼다.
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        x, y = len(nums1), len(nums2)
        low, high = 0, x

        while low <= high:
            partitionX = (low + high) // 2
            partitionY = (x + y + 1) // 2 - partitionX

            #num1 왼쪽 부분 배열에서 가장 큰 원소
            maxLeftX = float('-inf') if partitionX == 0 else nums1[partitionX - 1]
            #nuum1 오른쪽 부분 배열에서 가장 작은 원소
            minRightX = float('inf') if partitionX == x else nums1[partitionX]

            #num2 왼쪽 부분 배열에서 가장 큰 원소
            maxLeftY = float('-inf') if partitionY == 0 else nums2[partitionY - 1]
            #num2 오른쪽 부분 배열에서 가장 작은 원소
            minRightY = float('inf') if partitionY == y else nums2[partitionY]

            #분할 지점은 중앙값을 찾기에 적절한 위치인지 확인.
            if maxLeftX <= minRightY and maxLeftY <= minRightX:
                if (x + y) % 2 == 0:
                    return (max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2
                else:
                    return max(maxLeftX, maxLeftY)
            #아니면 포인터를 옮긴다.
            elif maxLeftX > minRightY:
                high = partitionX - 1
            else:
                low = partitionX + 1
